index.bs

```diff
@@ -68,12 +71,145 @@ The appropriate times to <a spec=HTML>fetch and process the linked resource</a>
+  This reduces the risk of unintended activity by user agents which are unaware of most recently added directives which might limit the scope of a rule.
+
+<div algorithm="parse speculation rules">
+  To <dfn>parse speculation rules</dfn> given a [=string=] |input| and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule set=] or null.
+
+  1. Let |parsed| be the result of [=Parse JSON into Infra values|parsing JSON into Infra values=] given |input|.
+  1. If |parsed| is not a [=map=], then return null.
+  1. Let |result| be an empty [=speculation rule set=].
+  1. If |parsed|["`prefetch`"] [=map/exists=] and is a [=list=], then [=list/for each=] |prefetchRule| of |parsed|["`prefetch`"]:
+    1. If |prefetchRule| is not a [=map=], then [=iteration/continue=].
+    1. Let |rule| be the result of [=parsing a speculation rule=] given |prefetchRule| and |baseURL|.
+    1. If |rule| is null, then [=iteration/continue=].
+    1. [=list/Append=] |rule| to |result|'s [=speculation rule set/prefetch rules=].
+  1. Return |result|.
+</div>
+
+<div algorithm="parse a speculation rule">
+  To <dfn>parse a speculation rule</dfn> given a [=map=] |input| and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule=] or null.
+
+  1. If |input| has any [=map/key=] other than "`source`", "`urls`", and "`requires`", then return null.
+  1. If |input|["`source`"] does not [=map/exist=] or is not the [=string=] "`list`", then return null.
+  1. Let |urls| be an empty [=list=].
+  1. If |input|["`urls`"] does not [=map/exist=], is not a [=list=], or has any element which is not a [=string=], then return null.
+  1. [=list/For each=] |urlString| of |input|["`urls`"]:
+    1. Let |parsedURL| be the result of [=basic URL parser|parsing=] |urlString| with |baseURL|.
+    1. If |parsedURL| is failure, then [=iteration/continue=].
+    1. If |parsedURL|'s [=url/scheme=] is not an [=HTTP(S) scheme=], then [=iteration/continue=].
+    1. [=list/Append=] |parsedURL| to |urls|.
+  1. Let |requirements| be an empty [=ordered set=].
+  1. If |input|["`requires`"] [=map/exists=], but is not a [=list=], then return null.
+  1. [=list/For each=] |requirement| of |input|["`requires`"]:
+    1. If |requirement| is not the [=string=] "`anonymous-client-ip`", then return null.
+    1. [=set/Append=] |requirement| to |requirements|.
+  1. Return a [=speculation rule=] with [=speculation rule/URLs=] |urls| and [=speculation rule/requirements=] |requirements|.
+</div>
+
+<h3 id="speculation-rules-processing">Processing Model</h3>
+
+A [=document=] has a <dfn for=document export>list of speculation rule sets</dfn>, which is an initially empty [=list=].
+
+<!-- TODO(domfarolino): Get rid of the `data-link-type="interface"` once we fix the dfn in HTML. -->
+Periodically, the user agent may [=queue a task=] to the <a data-link-type="interface">DOM manipulation task source</a> to [=consider speculation=] for any [=document=].
```

Comment:
```suggestion
Periodically, for any [=document=] |document|, the user agent may [=queue a global task=] on the <a data-link-type="interface">DOM manipulation task source</a> with |document|'s [=relevant global object=] to [=consider speculation=] for |document|.
```

Maybe we should require that _document_ is fully active?

Reply:
Added a check within the task (because this could technically change in between, and nothing stops the UA from lifting that check for efficiency). I think prerendered pages (and possibly also bfcache?) still qualify as fully active, so I've left an inline issue to refine that.